src/app/content/utils/canonicalUrl.ts

```diff
@@ -14,29 +15,50 @@ export async function getCanonicalUrlParams(
   pageId: string,
   bookVersion?: string
 ) {
-  const bookDefaultMap = [[book.id, {}]] as Array<[string, ObjectLiteral<undefined>]>;
   const getBook = makeUnifiedBookLoader(archiveLoader, osWebLoader);
-  const bookVersionFromConfig = getBookVersionFromUUIDSync(book.id);
-  const canonicals = ([
-    ...(CANONICAL_MAP[book.id] || []),
-    ...(bookVersionFromConfig && bookVersion === bookVersionFromConfig.defaultVersion ? bookDefaultMap : []),
-    // use the current book as a last resort if it has the same version as in books config
-  ]).filter(([id]) => !!getBookVersionFromUUIDSync(id));
 
-  for (const [id, CANONICAL_PAGES_MAP] of canonicals) {
-    const version = assertDefined(
-      getBookVersionFromUUIDSync(id),
-      `We've already filtered out books that are not in the BOOK configuration`
-    ).defaultVersion;
-    const canonicalBook = book.id === id  && hasOSWebData(book) ? book : await getBook(id, version);
-    const mappedPageId = CANONICAL_PAGES_MAP[pageId] || pageId;
-    const treeSection = findArchiveTreeNodeById(canonicalBook.tree, mappedPageId);
+  const getCanonicalMap = (bookId: string) => {
+    const bookDefaultMap = [[bookId, {}]] as Array<[string, ObjectLiteral<undefined>]>;
+    const bookVersionFromConfig = getBookVersionFromUUIDSync(bookId);
+    return ([
+      ...(CANONICAL_MAP[bookId] || []),
+      ...(bookVersionFromConfig && bookVersion === bookVersionFromConfig.defaultVersion ? bookDefaultMap : []),
+      // use the current book as a last resort if it has the same version as in books config
+    ]).filter(([id]) => !!getBookVersionFromUUIDSync(id));
+};
 
-    if (treeSection) {
-      const pageInBook = assertDefined(treeSection.slug, 'Expected page to have slug.');
-      return {book: {slug: canonicalBook.slug}, page: {slug: pageInBook}};
+  let canonicalMap = getCanonicalMap(book.id);
+  let canonicalPageId = pageId;
+  let done = false;
+  let canonicalBook;
+  let treeSection;
+
+  while (!done) {
+    for (const [id, CANONICAL_PAGES_MAP] of canonicalMap) {
+      const version = assertDefined(
+        getBookVersionFromUUIDSync(id),
+        `We've already filtered out books that are not in the BOOK configuration`
+      ).defaultVersion;
+      const useCurrentBookAsCanonical = book.id === id  && hasOSWebData(book);
```

Comment:
Might have to check that the book.version also matches the default version?

Reply:
Hmm thinking through this... If we add this check and they don't match, then on line 43 we'll get the latest version of the book and check for the current page in that book's tree. If the page isn't in the tree then no canonical URL is used. I think that's probably what we want? What do you think?

Reply:
I think with the current code if we got into a situation where the book is in its own `CANONICAL_MAP[bookId]` (will this happen?) then this check would be true for any version of the book so they would all have themselves as canonical links. This doesn't happen with the bookDefaultMap simply because the bookDefaultMap is not added if the version doesn't match.

I think the desired behavior in the case you described would be to find the latest version of the book that has the removed page, which would be a pain...

Reply:
I think it would be a good idea to add a few more tests with the expected behavior for some of the different cases so we can make sure they all work:

- Book has no CANONICAL_MAP and is defaultVersion
- Book has no CANONICAL_MAP and is not defaultVersion. Page is in the defaultVersion
- Book has no CANONICAL_MAP and is not defaultVersion. Page is missing from the defaultVersion
- Book has itself as CANONICAL_MAP and is defaultVersion
- Book has itself as CANONICAL_MAP and is not defaultVersion. Page is in the map.
- Book has itself as CANONICAL_MAP and is not defaultVersion. Page is missing from the map.
- Book has another book as CANONICAL_MAP and is defaultVersion
- Book has another book as CANONICAL_MAP and is not defaultVersion. Page is in the map.
- Book has another book as CANONICAL_MAP and is not defaultVersion. Page is missing from the map.

Reply:
I've updated the specs. Did not need to add much since we decided we're not concerned with loaded page version.